src/Components/Web/test/Image/ImageTest.cs

```diff
@@ -0,0 +1,254 @@
+// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+
+#nullable enable
+
+using System;
+using System.Collections.Generic;
+using System.Collections.Concurrent;
+using System.Diagnostics;
+using System.Threading;
+using System.Threading.Tasks;
+using System.Linq;
+using System.Text.Json;
+using System.IO;
+using Microsoft.AspNetCore.Components.Test.Helpers;
+using Microsoft.Extensions.DependencyInjection;
+using Microsoft.JSInterop;
+using Xunit;
+using Microsoft.AspNetCore.Components;
+using Microsoft.AspNetCore.Components.RenderTree;
+
+namespace Microsoft.AspNetCore.Components.Web.Image.Tests;
+
+/// <summary>
+/// Unit tests for the Image component
+/// </summary>
+public class ImageTest
+{
+    private static readonly byte[] PngBytes = Convert.FromBase64String("iVBORw0KGgoAAAANSUhEUgAAAAEAAAABCAYAAAAfFcSJAAAAAXNSR0IArs4c6QAAAA1JREFUGFdjqK6u/g8ABVcCcYoGhmwAAAAASUVORK5CYII=");
+
+    [Fact]
+    public async Task LoadsImage_InvokesSetImageAsync_WhenSourceProvided()
+    {
+        var js = new FakeImageJsRuntime(cacheHit: false);
+        using var renderer = CreateRenderer(js);
+        var comp = (Image)renderer.InstantiateComponent<Image>();
+        var id = renderer.AssignRootComponentId(comp);
+
+        var source = new ImageSource(PngBytes, "image/png", cacheKey: "png-1");
+        await renderer.RenderRootComponentAsync(id, ParameterView.FromDictionary(new Dictionary<string, object?>
+        {
+            [nameof(Image.Source)] = source,
+        }));
+
+        Assert.Equal(1, js.Count("Blazor._internal.BinaryImageComponent.setImageAsync"));
+    }
+
+    [Fact]
+    public async Task SkipsReload_OnSameCacheKey()
+    {
+        var js = new FakeImageJsRuntime(cacheHit: false);
+        using var renderer = CreateRenderer(js);
+        var comp = (Image)renderer.InstantiateComponent<Image>();
+        var id = renderer.AssignRootComponentId(comp);
+
+        var s1 = new ImageSource(new byte[10], "image/png", cacheKey: "same");
+        await renderer.RenderRootComponentAsync(id, ParameterView.FromDictionary(new Dictionary<string, object?>
+        {
+            [nameof(Image.Source)] = s1,
+        }));
+
+        var s2 = new ImageSource(new byte[20], "image/png", cacheKey: "same");
+        await renderer.RenderRootComponentAsync(id, ParameterView.FromDictionary(new Dictionary<string, object?>
+        {
+            [nameof(Image.Source)] = s2,
+        }));
+
+        // Implementation skips reloading when cache key unchanged.
+        Assert.Equal(1, js.Count("Blazor._internal.BinaryImageComponent.setImageAsync"));
+    }
+
+    [Fact]
+    public async Task NullSource_Throws()
```

Comment:
What will happen if `src` points to non-image MIME, `.txt`, `.pdf`, `.json`? What if image data is corrupted? Can we display e.g. `.tiff`?

Reply:
> Thanks @javiercn Any functionality for thumbnails in case there was an image error ?

We haven't done anything explicitly yet, but it is in our minds. The scenario of being able to provide something like a blur hash or similar is something we want to support (assuming for example that you had a pre-computed data url for the image stored in the DB or similar). 

That said, I believe that you should be able to provide a `src` element on the initial load. We would render that and naturally replace it with the obj url once we've downloaded the stream.

Reply:
> Is it any different compared to if you did the same with a regular `img` tag? The browser will simply avoid loading the image, raise `onerror` and display the `alt` text inside the box? (I'm guessing)

I think it's worth having a test for such scenarios.